### PR TITLE
Extract resolver tests to their own crate 

### PR DESCRIPTION
**Pull request by alexcrichton**
These tests take a good amount of time to run locally and they're also
causing a lot of dependencies to get pulled into rust-lang/rust, so
let's have a separate crate that we just test on our own CI

### CODE REVIEW

**Comment by Eh2406** [comment]
I have 2 open resolver related PRs and hopefully an additional [one](https://github.com/rust-lang/cargo/compare/master...Eh2406:public_dependency-as-type) on the way, all of witch will need to be merged. If this becomes urgent we should land it! If we have some time I'd like to think some more. 

The resolver test have different dependencies, tooling, and structure. I like that this gives us a place to  make that cleare. This also makes it much lower cost to make them even more different. Like adding benchmarks.

I have some ergonomics concerns. How easy is it to get Jetbrains/RLS to run the tests in there new location? Does that involve an additional copy of `/target` for all of cargos dependencies? Does this help with the build time for resolver tests? Does this help with the build time for non resolver tests? Do the debug assertions work correctly?

More generally I wonder if this goes far enough. I have for a while thought that we have too many resolver tests. I think a lot of the property based tests are approximations of the new SAT bass test. When we have some experience with the new test, I would strongly consider just removing the other ones. All the example bass test can be made to be of the same form: does some dependencies resolved against a specified index. We could store the outside indexes in a folder of Json or something, and let the SAT framework tell us whether we got the right answer. We could go a step further and just have it output the lock file or error message, and snapshot it. These changes would make it even more true that's a resolver has a different test infrastructure and strategy then everything else.

The thing that would make running the resolver tests much more pleasant is making the resolver a separate crate that does not depend on cargo. The only way I see making that happen is making a trate for each type that cargo gives to the resolver. Then making `core/resolver` a series of `impl ReloverX for X` and one call to a library function. But this would be a project.

**Comment by alexcrichton** [comment]
This isn't particularly pressing, no, I was largely just interested in getting updates to rust-lang/rust working again which @ehuss mentioned is blocked on `smallvec` due to a compiler bug, but that compiler bug may have been fixed by this point as well (and needs to be fixed regardless).

For your questions:

> How easy is it to get Jetbrains/RLS to run the tests in there new location?

Without being a user of these extensions, I'm not sure! 

> Does that involve an additional copy of /target for all of cargos dependencies?

Currently, yes, because the integration in rust-lang/rust makes using `[workspace]` particularly tricky and I don't think we can do it just yet.

> Does this help with the build time for resolver tests?

Presumably yes because there's less code to compile!

> Does this help with the build time for non resolver tests?

Probably not since this is just a few test being extracted.

> Do the debug assertions work correctly?

I believe they should as usual, yes.

> The thing that would make running the resolver tests much more pleasant is making the resolver a separate crate that does not depend on cargo

Agreed! You're right in that this'll be super tricky, but it may also be possible (and a good idea) to separate out the core types of Cargo like `Summary` and such. That would, I think, basically allow the resolver to stand on its own which would be a good idea for sure.

That's certainly a larger undertaking though :)

File: crates/resolver-tests/src/lib.rs [hunk 1]
```diff
@@ -182,7 +178,7 @@ pub fn resolve_with_config_raw(
 
     // The largest test in our suite takes less then 30 sec.
     // So lets fail the test if we have ben running for two long.
-    assert!(start.elapsed() < slow_cpu_multiplier(60));
+    assert!(start.elapsed().as_secs() < 60);
```
Comment (Eh2406) [inline crates/resolver-tests/src/lib.rs]: Why remove `slow_cpu_multiplier`? Do we think that the other people (#6491) running Cargos CI will not run these tests?

**Comment by Eh2406** [comment]
To be clear, with an answer to my nit question and a rebase I think this should be accepted at the earliest convenience. If my PRs ever get in good enough shape to be worth merging, I will happily rebase.

File: crates/resolver-tests/src/lib.rs (hunk 1 above)
Comment (alexcrichton) [reply crates/resolver-tests/src/lib.rs]: This was just a big chunk of code to migrate that didn't really seem worth it. The workaround can always be re-added if still necessary.

**Comment by alexcrichton** [comment]
Oh sorry I just ran out of time to rebase this and get it through, I will do so now.

**Comment by Eh2406** [comment]
> Oh sorry I just ran out of time

It has been clear that you have been very busy for an extended period of time, working on many divers projects each important to keeping this Rust experiment going.
I just wanted to interrupt your todo list to say, Thank you for all the work! Thank you for keeping things going!
I know it is not a replacement for more time nor for more help, but it is what I can do.

**Pull request merged**